types/lodash/lodash-tests.ts

```diff
@@ -7389,3 +7407,152 @@ _.templateSettings; // $ExpectType TemplateSettings
+        [sym]: 'sym-OK';
+        [key: string]: 'OK';
+    }
+
+    interface O2 { [key: string]: 'OK' };
+
+    type O3 = Record<string | typeof sym, 'OK'>;
+
+    {   // Trivial cases:
+        type A = GetIndexedField<AbcObject, 'a'>; // $ExpectType number
+        type B = GetIndexedField<AbcObject, 'b'>; // $ExpectType string
+        type D = GetIndexedField<AbcObject, '1'>; // $ExpectType undefined
+        type F = GetIndexedField<AbcObject, number>; // $ExpectType undefined
+        type G = GetIndexedField<O1, typeof sym>; // $ExpectType "sym-OK"
+        type H = GetIndexedField<O1, string>; // $ExpectType "OK"
+        type I = GetIndexedField<O3, typeof sym>; // $ExpectType "OK"
+    }
+
+    {   // Note, TS resolves keyof Ox differently for our test object types:
+        type A = keyof O1 // $ExpectType keyof O1
+        type B = keyof O2 // $ExpectType number | string
+        type C = keyof O3 // $ExpectType string | typeof sym
+
+        // However, in all cases TS knows a number type can be used for indexing:
+        type D = O1[number]; // $ExpectType "OK"
+        type E = O2[number]; // $ExpectType "OK"
+        type F = O3[number]; // $ExpectType "OK"
+
+        // And GetIndexedField<> should behave the same:
+        type G = GetIndexedField<O1, number>; // $ExpectType "OK"
+        type H = GetIndexedField<O2, number>; // $ExpectType "OK"
+        type I = GetIndexedField<O3, number>; // $ExpectType "OK"
+        type J = GetIndexedField<O3, 1>; // $ExpectType "OK"
+    }
+
+    {
+        type E = GetIndexedField<O1, bigint>; // $ExpectType undefined
+        type F = GetIndexedField<O2, boolean>; // $ExpectType undefined
+        type G = GetIndexedField<O3, null>; // $ExpectType undefined
+        type H = GetIndexedField<O1, undefined>; // $ExpectType undefined
+    }
+}
+
+// GetIndexedField: indexing strings
+
+{
+    type S1 = string;
+    type S2 = string & { x: 'OK' };
+
+    type A = GetIndexedField<S1, number>; // $ExpectType string
+    type B = GetIndexedField<S1, 1>; // $ExpectType string
+    type C = GetIndexedField<S1, '1'>; // $ExpectType string
+    type D = GetIndexedField<S1, 'length'>; // $ExpectType number
+    type F = GetIndexedField<S1, 'x'>; // $ExpectType undefined
+    type G = GetIndexedField<S2, 'x'>; // $ExpectType "OK"
+}
+
+// GetIndexedFields: indexing tuples
+
+{
+    type T1 = ['OK'];
+
+    type A = GetIndexedField<T1, number>; // $ExpectType "OK"
+    type B = GetIndexedField<T1, 0>; // $ExpectType "OK"
+    type C = GetIndexedField<T1, 1>; // $ExpectType undefined
+    type D = GetIndexedField<T1, '0'>; // $ExpectType "OK"
+    type E = GetIndexedField<T1, '1'>; // $ExpectType undefined
+    type F = GetIndexedField<T1, 'length'>; // $ExpectType 1
+}
+
+// GetFieldType: correctly handles any possible combination of dots and square
+// brackets in path type literals.
+{
+    interface O1 { key: 'value' };
+    interface O2 { some: { test: { path: 'value' } } };
```

Comment:
The name O1/O2/O3 isn't very descriptive. If possible, can you name it based on its contents? E.g. NestedObject.

I'm not sure if that's possible for all of these, so another option is to move the definitions of O2 and O3 down closer to where they are used, so the reader doesn't need to search as far to find the definition.